File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/dc37cd46e578d29248104389008598aa.go
package main

import (
	"fmt"
	"sort"
)

type pii struct {
	first  int
	second int
}

type F struct {
	p int
	q int
}

type Fenw struct {
	tr [202020]int
}

func (f *Fenw) add(pos int) {
	for i := pos; i <= 200020; i |= i + 1 {
		f.tr[i]++
	}
}

func (f *Fenw) sum(pos int) int {
	res := 0
	for i := pos; i >= 0; i = (i & (i + 1)) - 1 {
		res += f.tr[i]
	}
	return res
}

func (f *Fenw) rangeSum(l, r int) int {
	return f.sum(r) - f.sum(l-1)
}

func solve1(v []pii, w int) int64 {
	allF := make([]F, 0)
	for _, p := range v {
		from := F{p.first, p.second + w}
		to := F{p.first, p.second - w}
		allF = append(allF, from)
		allF = append(allF, to)
	}
	allF = append(allF, F{0, 1})
	sort.Slice(allF, func(i, j int) bool {
		return allF[i].p*int64(allF[j].q) < allF[j].p*int64(allF[i].q)
	})
	allF = unique(allF)
	for i, p := range v {
		from := F{p.first, p.second + w}
		to := F{p.first, p.second - w}
		v[i] = pii{lowerBound(allF, from), lowerBound(allF, to)}
	}
	sort.Slice(v, func(i, j int) bool {
		return v[i].first < v[j].first || (v[i].first == v[j].first && v[i].second > v[j].second)
	})
	var res int64
	f1 := Fenw{}
	for i := 0; i < len(v); i++ {
		res += int64(i - f1.sum(v[i].second-1))
		f1.add(v[i].second)
	}
	return res
}

func solve2(v [2][]pii, w int) int64 {
	allF := make([]F, 0)
	for _, p := range v {
		for _, q := range p {
			from := F{q.first, q.second + w}
			to := F{q.first, q.second - w}
			allF = append(allF, from)
			allF = append(allF, to)
		}
	}
	allF = append(allF, F{0, 1})
	sort.Slice(allF, func(i, j int) bool {
		return allF[i].p*int64(allF[j].q) < allF[j].p*int64(allF[i].q)
	})
	allF = unique(allF)
	var res int64
	f := [2]Fenw{}
	for i := 0; i < 2; i++ {
		for j := 0; j < len(v[i]); j++ {
			from := F{v[i][j].first, v[i][j].second + w}
			to := F{v[i][j].first, v[i][j].second - w}
			v[i][j] = pii{lowerBound(allF, from), lowerBound(allF, to)}
			f[i].add(v[i][j].first)
		}
	}
	for _, p := range v[0] {
		res += int64(f[1].rangeSum(p.first, p.second))
	}
	for _, p := range v[1] {
		res += int64(f[0].rangeSum(p.first+1, p.second))
	}
	return res
}

func lowerBound(arr []F, target F) int {
	left, right := 0, len(arr)
	for left < right {
		mid := left + (right-left)/2
		if arr[mid].p < target.p || (arr[mid].p == target.p && arr[mid].q < target.q) {
			left = mid + 1
		} else {
			right = mid
		}
	}
	return left
}

func unique(arr []F) []F {
	j := 0
	for i := 1; i < len(arr); i++ {
		if arr[i] != arr[j] {
			j++
			arr[j] = arr[i]
		}
	}
	return arr[:j+1]
}

func main() {
	var n, w int
	fmt.Scan(&n, &w)
	x := make([]int, n)
	a := make([]int, n)
	v := [2][]pii{}
	for i := 0; i < n; i++ {
		fmt.Scan(&x[i], &a[i])
		if x[i] < 0 {
			v[0] = append(v[0], pii{abs(x[i]), abs(a[i])})
		} else {
			v[1] = append(v[1], pii{abs(x[i]), abs(a[i])})
		}
	}
	ans := solve1(v[0], w) + solve1(v[1], w) + solve2(v, w)
	fmt.Println(ans)
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Java/dc37cd46e578d29248104389008598aa.java
import java.util.*;

class Main {
    static class Pair {
        int first;
        int second;

        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }

    static class F implements Comparable<F> {
        int p;
        int q;

        F(int p, int q) {
            this.p = p;
            this.q = q;
        }

        @Override
        public int compareTo(F f) {
            return Integer.compare(this.p * f.q, f.p * this.q);
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            }
            if (obj == null || getClass() != obj.getClass()) {
                return false;
            }
            F f = (F) obj;
            return this.p * f.q == f.p * this.q;
        }
    }

    static class Fenw {
        int[] tr;

        Fenw() {
            tr = new int[202020];
        }

        void add(int pos) {
            for (int i = pos; i <= 200020; i |= i + 1) {
                tr[i]++;
            }
        }

        int sum(int pos) {
            int res = 0;
            for (int i = pos; i >= 0; i = (i & (i + 1)) - 1) {
                res += tr[i];
            }
            return res;
        }

        int sum(int l, int r) {
            return sum(r) - sum(l - 1);
        }
    }

    static Fenw f1 = new Fenw();
    static Fenw[] f = { new Fenw(), new Fenw() };

    static long solve1(List<Pair> v) {
        List<F> allF = new ArrayList<>();
        for (Pair p : v) {
            F from = new F(p.first, p.second + w);
            F to = new F(p.first, p.second - w);
            allF.add(from);
            allF.add(to);
        }
        allF.add(new F(0, 1));
        Collections.sort(allF);
        allF = reunique(allF);
        for (Pair p : v) {
            F from = new F(p.first, p.second + w);
            F to = new F(p.first, p.second - w);
            p.first = lowerBound(allF, from);
            p.second = lowerBound(allF, to);
        }
        Collections.sort(v, (p1, p2) -> {
            if (p1.first != p2.first) {
                return Integer.compare(p1.first, p2.first);
            } else {
                return Integer.compare(p1.second, p2.second);
            }
        });
        long res = 0;
        Arrays.fill(f1.tr, 0);
        for (int i = 0; i < v.size(); ++i) {
            res += i - f1.sum(v.get(i).second - 1);
            f1.add(v.get(i).second);
        }
        return res;
    }

    static long solve2(List<Pair>[] v) {
        List<F> allF = new ArrayList<>();
        for (int i = 0; i < 2; i++) {
            for (Pair p : v[i]) {
                F from = new F(p.first, p.second + w);
                F to = new F(p.first, p.second - w);
                allF.add(from);
                allF.add(to);
            }
        }
        allF.add(new F(0, 1));
        Collections.sort(allF);
        allF = reunique(allF);
        long res = 0;
        List<Triple> events = new ArrayList<>();
        for (int i = 0; i < 2; i++) {
            for (Pair p : v[i]) {
                F from = new F(p.first, p.second + w);
                F to = new F(p.first, p.second - w);
                p.first = lowerBound(allF, from);
                p.second = lowerBound(allF, to);
                f[i].add(p.first);
            }
        }

        for (Pair p : v[0]) {
            res += f[1].sum(p.first, p.second);
        }
        for (Pair p : v[1]) {
            res += f[0].sum(p.first + 1, p.second);
        }
        return res;
    }

    static int lowerBound(List<F> list, F target) {
        int low = 0;
        int high = list.size();
        while (low < high) {
            int mid = (low + high) / 2;
            if (list.get(mid).compareTo(target) < 0) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }

    static List<F> reunique(List<F> list) {
        List<F> result = new ArrayList<>();
        for (F f : list) {
            if (result.isEmpty() || !result.get(result.size() - 1).equals(f)) {
                result.add(f);
            }
        }
        return result;
    }

    static class Triple {
        int a;
        int b;
        int c;

        Triple(int a, int b, int c) {
            this.a = a;
            this.b = b;
            this.c = c;
        }
    }

    static int[] x;
    static int[] a;
    static List<Pair>[] v;
    static int w;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        w = sc.nextInt();
        x = new int[n];
        a = new int[n];
        v = new List[2];
        v[0] = new ArrayList<>();
        v[1] = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            x[i] = sc.nextInt();
            a[i] = sc.nextInt();
            v[x[i] < 0 ? 0 : 1].add(new Pair(Math.abs(x[i]), Math.abs(a[i])));
        }
        long ans = 0;
        ans += solve1(v[0]);
        ans += solve1(v[1]);
        ans += solve2(v);
        System.out.println(ans);
    }
}
